### ISSUE TITLE
composeUp using captureContainersOutputToFiles blocks gradle from completing

### ISSUE THREAD

**Issue by JWood48** [avast/gradle-docker-compose-plugin#188]
Using composeUp and configuring a output log directory like: 

captureContainersOutputToFiles = "..."

Prevents gradle from completing and gradle hangs with the following output:

BUILD SUCCESSFUL in 23s
1 actionable tasks: 1 executed
<-------------> 0% WAITING
> IDLE
> IDLE

I would expect the task to complete and the logs being written to the configured directory?

Or is it the gradle process that is piping the logs to the log directory so it must keep running?

(test using gradle 5.2.1)

**Comment by augi** [comment]
Could you please prepare a simple project that proves this behavior? Thanks!

**Comment by LukasWysocki** [comment]
Hit the same problem.
Repo with reproducer is here:
https://github.com/LukasWysocki/gradle-docker-compose-plugin-issue-188

This is docker-compose using busybox to run single command "sleep 42"

to test do:

```
git clone git@github.com:LukasWysocki/gradle-docker-compose-plugin-issue-188.git
cd gradle-docker-compose-plugin-issue-188
time gradle composeUp
```

The effect is:
```
lwysocki@lwy-thinkpad:~/high5/gradle-docker-compose-plugin-issue-188$ time gradle composeUp

> Task :composeUp
sleep uses an image, skipping
Will use 172.24.0.1 (network 844c3edbeb4315546c21b61404a6a645_gradle-docker-compose-plugin-issue-188__default) as host of sleep

BUILD SUCCESSFUL in 6s
1 actionable task: 1 executed

real    0m47,509s
user	0m12,744s
sys	0m0,704s

```

Notice `BUILD SUCCESSFUL in 6s` however time reports  `real    0m47,509s` . Which is the ~42 seconds of sleep + ~6s of build time.

**Comment by augi** [comment]
Thank you for your time you invested to reporting this issue! 👍 

Unfortunately, I'm not able to reproduce the issue even with repository you prepared :( My output is this:
```
$ time ./gradlew composeUp

> Task :composeUp
sleep uses an image, skipping
Will use localhost as host of sleep

BUILD SUCCESSFUL in 6s
1 actionable task: 1 executed

real    0m7.404s
user    0m0.045s
sys     0m0.197s
```

The only difference is that you are executing `gradle` command but I don't have Gradle locally installed so I'm using Gradle Wrapper.

The thread to read containers output is executed [as background thread](https://github.com/avast/gradle-docker-compose-plugin/blob/master/src/main/groovy/com/avast/gradle/dockercompose/ComposeExecutor.groovy#L113) so it shouldn't block process shutdown.

My only idea is that you are using [Gradle daemon](https://docs.gradle.org/current/userguide/gradle_daemon.html) - so could you please execute the same command with `--no-daemon --debug` added?

**Comment by augi** [comment]
To fix this issue, we should forcibly close the daemon threads when Gradle _execution_ finishes. Then using Gradle daemon, it's not the same as JVM shutdown. Unfortunately, I'm not aware of any Gradle API that would allow this.

On other hand, the current behavior is required for some use-cases, like people start using `composeUp`, do other things (outside of Gradle), and then stop the containers using `composeDown` (in separate Gradle execution).

I would like to add an option to change the behavior to support also your use-case anyway! If you know about a Gradle hook that would be executed after Gradle _execution_ finishes, please let me know.

**Comment by augi** [comment]
Now, we are using a hook to interrupt the daemon threads after Gradle build finishes. It's released in `0.9.3`.
